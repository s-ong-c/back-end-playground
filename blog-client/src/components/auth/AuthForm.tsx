import * as React from 'react';
import styled from 'styled-components';
import useInput from '../../lib/hooks/useInput';
import AuthEmailForm from './AuthEmailForm';
import { AuthMode } from '../../contexts/CoreContext';

const AuthFormBlock = styled.div``;
interface AuthFormProps{
    mode: AuthMode;
}

const AuthForm: React.SFC<AuthFormProps> = ({ mode }) => {
    const [email, onChangeEmail] = useInput('');
    const onSubmit = (email: string) => {
        console.log(email);
    };
    const modeText = mode === 'REGISTER' ? '회원가입' : '로그인';
    return (
        <AuthFormBlock>
            <h2>{modeText}</h2>
            <AuthEmailForm 
                value={email} 
                onChange={onChangeEmail} 
                onSubmit={onSubmit} 
                mode={mode}
            />
        </AuthFormBlock>
       );
};

export default AuthForm;