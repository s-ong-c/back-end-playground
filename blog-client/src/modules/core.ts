import { createStandardAction } from 'typesafe-actions';
import produce from 'immer';
import { createReducer } from '../lib/utils';

// actions 함수
const SET_LAYER = 'core/SET_LAYER';
const SHOW_AUTH_MODAL = 'core/SHOW_AUTH_MODAL';
const CHANGE_AUTH_MODAL_MODE = 'core/CHANGE_AUTH_MODAL_MODE';
const CLOSE_AUTH_MODAL = 'core/CLOSE_AUTH_MODAL';

export type AuthMode = 'REGISTER' | 'LOGIN';

export const setLayer = createStandardAction(SET_LAYER)<boolean>();
export const showAuthModal = createStandardAction(SHOW_AUTH_MODAL)<AuthMode>();
export const changeAuthModalMode = createStandardAction(CHANGE_AUTH_MODAL_MODE)<
  AuthMode
>();
export const closeAuthModal = createStandardAction(CLOSE_AUTH_MODAL)();

type SetLayer = ReturnType<typeof setLayer>;
type ShowAuthModal = ReturnType<typeof showAuthModal>;
type ChangeAuthModalMode = ReturnType<typeof changeAuthModalMode>;

export type CoreState = {
  layer: boolean;
  auth: {
    visible: boolean;
    mode: AuthMode;
  };
};
const initialState: CoreState = {
  layer: false,
  auth: {
    visible: false,
    mode: 'LOGIN',
  },
};
const core = createReducer<CoreState>(
  {
    [SET_LAYER]: (state, action: SetLayer) => ({
      ...state,
      layer: action.payload,
    }),
    [SHOW_AUTH_MODAL]: (state, action: ShowAuthModal) =>
      produce(state, draft => {
        draft.auth.mode = action.payload;
        draft.auth.visible = true;
        draft.layer = true;
      }),
    [CHANGE_AUTH_MODAL_MODE]: (state, action: ChangeAuthModalMode) =>
      produce(state, draft => {
        draft.auth.mode = action.payload;
      }),
    [CLOSE_AUTH_MODAL]: state =>
      produce(state, draft => {
        draft.auth.visible = false;
        draft.layer = false;
      }),
  },
  initialState,
);

export default core;
