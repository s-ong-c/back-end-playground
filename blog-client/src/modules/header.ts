import { createReducer, updateKey } from '../lib/utils';
import { createStandardAction } from 'typesafe-actions';

const SET_CUSTOM = 'header/SET_CUSTOM';
const SET_USER_LOGO = 'header/SET_USER_LOGO';

type UserLogo = {
  title: string | null;
  logo_image: string | null;
};

export interface HeaderState {
  custom: boolean;
  userLogo: null | UserLogo;
}

export const setCustom = createStandardAction(SET_CUSTOM)<boolean>();
export const setUserLogo = createStandardAction(SET_USER_LOGO)<UserLogo>();

type SetCustom = ReturnType<typeof setCustom>;
type SetUserLogo = ReturnType<typeof setUserLogo>;

const initialState: HeaderState = {
  custom: false,
  userLogo: null,
};
const header = createReducer(
  {
    [SET_CUSTOM]: (state, action: SetCustom) => {
      return updateKey(state, 'custom', action.payload);
    },
    [SET_USER_LOGO]: (state, { payload }: SetUserLogo) => {
      return updateKey(state, 'userLogo', payload);
    },
  },
  initialState,
);

export default header;
