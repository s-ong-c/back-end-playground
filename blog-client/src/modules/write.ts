import { createStandardAction } from 'typesafe-actions';
import { createReducer, updateKey } from '../lib/utils';

const CHANGE_MARKDOWN = 'write/CHANGE_MARKDOWN';
const CHANGE_TITLE = 'write/CHANGE_TITLE';
const CONVERT_EDITOR_MODE = 'write/CONVERT_EDITOR_MODE';
const SET_HTML = 'write/SET_HTML';
const SET_TEXT_BODY = 'write/SET_TEXT_BODY';
const CHANGE_TAGS = 'write/CHANGE_TAGS';
const OPEN_PUBLISH = 'write/OPEN_PUBLISH';
const CLOSE_PUBLISH = 'write/CLOSE_PUBLISH';
const SET_DEFAULT_DESCRIPTION = 'write/SET_DEFAULT_DESCRIPTION';
const CHANGE_DESCRIPTION = 'write/CHANGE_DESCRIPTION';

export const changeMarkdown = createStandardAction(CHANGE_MARKDOWN)<string>();
export const changeTitle = createStandardAction(CHANGE_TITLE)<string>();
export const convertEditorMode = createStandardAction(CONVERT_EDITOR_MODE)();
export const setHtml = createStandardAction(SET_HTML)<string>();
export const changeTags = createStandardAction(CHANGE_TAGS)<string[]>();
export const openPublish = createStandardAction(OPEN_PUBLISH)<void>();
export const closePublish = createStandardAction(CLOSE_PUBLISH)<void>();
export const setTextBody = createStandardAction(SET_TEXT_BODY)<string>();
export const setDefaultDescription = createStandardAction(
  SET_DEFAULT_DESCRIPTION,
)<string>();
export const changeDescription = createStandardAction(CHANGE_DESCRIPTION)<
  string
>();

type ChangeMarkdown = ReturnType<typeof changeMarkdown>;
type ChangeTitle = ReturnType<typeof changeTitle>;
type SetHtml = ReturnType<typeof setHtml>;
type ChangeTags = ReturnType<typeof changeTags>;
type OpenPublish = ReturnType<typeof openPublish>;
type ClosePublish = ReturnType<typeof closePublish>;
type SetTextBody = ReturnType<typeof setTextBody>;
type SetDefaultDescription = ReturnType<typeof setDefaultDescription>;
type ChangeDescription = ReturnType<typeof changeDescription>;

export enum WriteMode {
  MARKDOWN = 'MARKDOWN',
  WYSIWYG = 'WYSIWYG',
}

export type WriteState = {
  mode: WriteMode;
  markdown: string;
  title: string;
  html: string;
  tags: string[];
  publish: boolean;
  textBody: string;
  defaultDescription: string;
  description: string;
};

const initialState: WriteState = {
  mode: WriteMode.WYSIWYG,
  markdown: '',
  title: '',
  html: '',
  tags: [],
  publish: false,
  textBody: '',
  defaultDescription: '',
  description: '',
};

const write = createReducer(
  {
    [CHANGE_MARKDOWN]: (state, action: ChangeMarkdown) =>
      updateKey(state, 'markdown', action.payload),
    [CHANGE_TITLE]: (state, action: ChangeTitle) =>
      updateKey(state, 'title', action.payload),
    [CONVERT_EDITOR_MODE]: state =>
      updateKey(
        state,
        'mode',
        state.mode === WriteMode.MARKDOWN
          ? WriteMode.WYSIWYG
          : WriteMode.MARKDOWN,
      ),
    [SET_HTML]: (state, action: SetHtml) =>
      updateKey(state, 'html', action.payload),
    [CHANGE_TAGS]: (state, action: ChangeTags) =>
      updateKey(state, 'tags', action.payload),
    [OPEN_PUBLISH]: state => updateKey(state, 'publish', true),
    [CLOSE_PUBLISH]: state => updateKey(state, 'publish', false),
    [SET_TEXT_BODY]: (state, { payload: textBody }: SetTextBody) =>
      updateKey(state, 'textBody', textBody),
    [SET_DEFAULT_DESCRIPTION]: (
      state,
      { payload: defaultDescription }: SetDefaultDescription,
    ) => updateKey(state, 'defaultDescription', defaultDescription),
    [CHANGE_DESCRIPTION]: (
      state,
      { payload: description }: ChangeDescription,
    ) => updateKey(state, 'description', description),
  },
  initialState,
);

export default write;
