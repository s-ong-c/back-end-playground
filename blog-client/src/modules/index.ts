import { combineReducers } from 'redux';
import core, { CoreState } from './core';
import write, { WriteState } from './write';
import header, { HeaderState } from './header';

export type RootState = {
  core: CoreState;
  write: WriteState;
  header: HeaderState;
};

const rootReducer = combineReducers({
  core,
  write,
  header,
});

export default rootReducer;
