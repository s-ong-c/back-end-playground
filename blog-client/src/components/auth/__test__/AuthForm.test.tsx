import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import AuthForm, { AuthFormProps } from '../AuthForm';
import { AuthMode } from '../../../modules/core';

describe('AuthForm', () => {
    const setup = (props: Partial<AuthFormProps> = {}) => {
       const initialProps: AuthFormProps = {
          mode:'REGISTER',
          onToggleMode:() => {},
          onSendAuthEmail: (email: string) => {},
          load: false,
          registered: null,
       };
     return render(<AuthForm {...initialProps} {...props} />);
    }
    it('renders correctly', () => {
      const { container } = setup();
     // expect(container).toMatchSnapshot();
    });

  describe('handles modes', () => {
    it('REGISTER', () => {
      const { getByText } = setup();
      getByText('회원가입');
    });
    it('LOGIN', () => {
      const { getByText } =setup({mode: 'LOGIN'});
      getByText('로그인');
    });
  });

  it('calls onToggleMode', () => {
    const onToggleMode = jest.fn();
    const { container } = setup({
      mode: 'REGISTER',
      onToggleMode,
    });
    const anchor = container.querySelector('a');
    expect(anchor).toBeTruthy();
    if (!anchor) return;
    fireEvent.click(anchor);
    expect(onToggleMode.mock.calls.length).toBe(1);
  });
});