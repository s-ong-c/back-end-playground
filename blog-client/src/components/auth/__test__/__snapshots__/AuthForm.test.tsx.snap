// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm renders correctly 1`] = `
<div>
  <div
    class="sc-bxivhb ctZEFR"
  >
    <div>
      <h2
        data-testid="title"
      >
        회원가입
      </h2>
      <section>
        <h4>
          이메일로 
          회원가입
        </h4>
        <form
          class="sc-bdVaJa kaHBVA"
        >
          <input
            placeholder="이메일을 입력하세요."
            tabindex="2"
            value=""
          />
          <button
            tabindex="3"
          >
            회원가입
          </button>
        </form>
      </section>
      <section>
        <h4>
          소셜 계정으로 
          회원가입
        </h4>
        <div
          class="sc-htpNat gNrJqv"
        >
          <button
            class="sc-bwzfXH ggDXRg"
            style="background: rgb(39, 46, 51);"
            tabindex="4"
          >
            <svg>
              github.svg
            </svg>
          </button>
          <button
            class="sc-bwzfXH fSrJCx"
            style="background: white;"
            tabindex="5"
          >
            <svg>
              google.svg
            </svg>
          </button>
          <button
            class="sc-bwzfXH ggDXRg"
            style="background: rgb(59, 89, 152);"
            tabindex="6"
          >
            <svg>
              facebook.svg
            </svg>
          </button>
        </div>
      </section>
    </div>
    <div
      class="foot"
    >
      <span>
        계정이 이미 있으신가요?
      </span>
      <a
        tabindex="7"
      >
        로그인
      </a>
    </div>
  </div>
</div>
`;
